File: src/PHP/gelu.php
<?php

use PHP2xAI\Runtime\PHP\Core\GraphRuntime;
use PHP2xAI\Tensor\Tensor;

include("../../vendor/autoload.php");

function runGelu(string $name, Tensor $input): void
{
	$input->setRequiresGrad(true);
	$output = $input->gelu();
	$runtime = GraphRuntime::createFromOutputTensor($output);
	$runtime->forward();
	$runtime->backward();
	$runtime->refreshTensorsData();

	echo "\n=== {$name} ===\n";
	$output->printData();
	$input->printGrad();
}

// Values around zero and in both GELU tails.
runGelu('Matrix: negative, zero and positive values', Tensor::createFromData([
	[-3.0, -1.0, 0.0],
	[1.0, 2.0, 3.0],
], 'matrix'));

// GELU is elementwise and works unchanged on token embeddings [B, L, D].
runGelu('Token embeddings', Tensor::createFromData([
	[[-2.0, -0.5], [0.5, 2.0]],
	[[-1.5, 0.0], [1.5, 3.0]],
], 'tokens'));

File: src/Torch/gelu.py
import torch
import torch.nn.functional as functional


def run_gelu(name: str, values: list) -> None:
    input_tensor = torch.tensor(values, dtype=torch.float32, requires_grad=True)
    output = functional.gelu(input_tensor, approximate="tanh")
    output.backward(torch.ones_like(output))

    print(f"\n=== {name} ===")
    print("output:")
    print(output)
    print("input grad:")
    print(input_tensor.grad)


# approximate="tanh" matches PHP2xAI's GELU implementation.
run_gelu(
    "Matrix: negative, zero and positive values",
    [[-3.0, -1.0, 0.0], [1.0, 2.0, 3.0]],
)
run_gelu(
    "Token embeddings",
    [[[-2.0, -0.5], [0.5, 2.0]], [[-1.5, 0.0], [1.5, 3.0]]],
)
